### PR TITLE
 Fix switching Priority Mode with an active draft message from crashing app

### PR DESCRIPTION
**Pull request by ctkochan22**
@Expensify/pullerbear 
cc @marcaaron 

### Details
This was a tricky bug to find. I'm mostly worried about regression so ideally we can really test this thoroughly

### Fixed Issues
$ https://github.com/Expensify/App/issues/7810
$ https://github.com/Expensify/App/issues/7811

### Tests / QA Steps
**Fix tests**
1. Go to Settings -> Preferences. Change the priority mode to "#focus". Verify that the LHN condenses
3. Then change the priority mode back to "Most Recent". Verify that the LHN expands
4. Now go back to the chat and type text in a chat. Verify that that user now has a draft "pencil" icon 
5. Go back to Settings -> Preferences. Change the priority mode to "#focus". Verify that the LHN condenses, and the user with the draft still has a pencil.
    - This is where the app crashes without this change
6. Change the priority mode to "Most Recent". Verify that the LHN expands 
APfitvvVAQ

**_Regression_**
Part of the bug was introduced in this PR (https://github.com/Expensify/App/pull/6084), so lets take their tests word for word and test them
1. Type text in the chat without sending it
    - Verify that you see the pencil next to the name and the order does not re-order
6. Type text in chat and send it
    - Verify that the pencil appears and disappears once its sent. The LHN order should change. User that you messaged should go to the top (but below pinned users)
8. Type text in he chat positioned in the middle. Then click on another contact.
    - The pencil icon should remain. User with an unsent draft should move to the top 
9. Go to the user that had the draft message and delete it.
    - Pencil icon should disappear and order should re-order
10. As a different user, message the first user. The message sender should move to the top
    - Verify that the message sender moves to the top

#### Contributor (PR Author) Checklist
- [x] I made sure to pull `main` before submitting my PR for review
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [ ] I clearly indicated the environment tests should be run in (Staging vs Production)
- [x] I wrote testing steps that cover success & fail scenarios (if applicable)
- [x] I ran the tests & they passed on **all platforms**
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified there are no console errors related to changes in this PR
- [ ] I followed proper code patterns (see [Reviewing the code](../PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I added comments when the code was not self explanatory
    - [ ] I put all copy / text shown in the product in all `src/languages/*` files (if applicable)
    - [ ] I followed proper naming convention for platform-specific files (if applicable)
    - [ ] I followed style guidelines (in [`Styling.md`](../STYLING.md)) for all style edits I made
- [ ] I followed the guidelines as stated in the [Review Guidelines](../PR_REVIEW_GUIDELINES.md)

#### PR Reviewer Checklist
- [ ] I verified the Author pulled `main` before submitting the PR
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the testing environment is mentioned in the test steps
- [ ] I verified testing steps cover success & fail scenarios (if applicable)
- [ ] I verified tests pass on **all platforms** & I tested again on all platforms
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified there are no console errors related to changes in this PR
- [ ] I verified proper code patterns were followed (see [Reviewing the code](../PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified comments were added when the code was not self explanatory
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files (if applicable)
    - [ ] I verified proper naming convention for platform-specific files was followed (if applicable)
    - [ ] I verified [style guidelines](../STYLING.md) were followed
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](../PR_REVIEW_GUIDELINES.md)

### Tested On

- [x] Web
- [x] Mobile Web
- [x] Desktop
- [x] iOS
- [x] Android

#### Web
Above

#### Desktop
https://recordit.co/hU5T21bVYm

#### iOS
https://recordit.co/sE7YS8nxIo

#### Android
https://recordit.co/2Tt6hj5jw2

### CODE REVIEW

**Comment by ctkochan22** [comment]
Actually, filter is not the solution. We use map because we need to replace old report object with new ones

File: src/pages/home/sidebar/SidebarLinks.js [hunk 1]
```diff
@@ -158,23 +159,34 @@ class SidebarLinks extends React.Component {
...
             currentlyViewedReportID: props.currentlyViewedReportID,
             orderedReports: [],
+            priorityMode: props.priorityMode,
             unreadReports: SidebarLinks.getUnreadReports(props.reports || {}),
         };
     }
 
     static getDerivedStateFromProps(nextProps, prevState) {
         const shouldReorder = SidebarLinks.shouldReorder(nextProps, prevState.orderedReports, prevState.currentlyViewedReportID, prevState.unreadReports);
+        const switchingPriorityModes = nextProps.priorityMode !== prevState.priorityMode;
```
Comment (marcaaron) [inline src/pages/home/sidebar/SidebarLinks.js:168]: Can we compare `this.props.priorityMode` with `nextProps.priorityMode` and leave `state` out of the equation entirely?

File: src/pages/home/sidebar/SidebarLinks.js [hunk 2]
```diff
@@ -158,23 +159,34 @@ class SidebarLinks extends React.Component {
...
+            // Then match and replace older reports with the newer report conversations from recentReports
+                .map(orderedReport => _.find(recentReports, recentReport => orderedReport.reportID === recentReport.reportID))
+
+            // Because we are using map, we have to filter out any undefined reports that may happen when switching priority modes
+                .filter(orderedReport => orderedReport !== undefined)
+                .value();
 
         return {
             orderedReports,
+            priorityMode: nextProps.priorityMode,
```
Comment (marcaaron) [inline src/pages/home/sidebar/SidebarLinks.js:189]: I _think_ we can remove this from `state` if we just look at current and next props only.

File: src/pages/home/sidebar/SidebarLinks.js [hunk 3]
```diff
@@ -158,23 +159,34 @@ class SidebarLinks extends React.Component {
...
+        const orderedReports = shouldReorder || switchingPriorityModes
             ? recentReports
-            : _.map(prevState.orderedReports,
-                orderedReport => _.chain(recentReports)
-                    .filter(recentReport => orderedReport.reportID === recentReport.reportID)
-                    .first()
-                    .value());
+            : _.chain(prevState.orderedReports)
+
+            // To preserve the order of the conversations, we map over the previous state's order of reports.
```
Comment (marcaaron) [inline src/pages/home/sidebar/SidebarLinks.js:178]: I probably have to study the changes in `SidebarLinks` some more, but any ideas why it is necessary to "preserve the order" of these conversations?

File: src/pages/home/sidebar/SidebarLinks.js [hunk 4]
```diff
@@ -158,23 +159,34 @@ class SidebarLinks extends React.Component {
...
+            // Then match and replace older reports with the newer report conversations from recentReports
+                .map(orderedReport => _.find(recentReports, recentReport => orderedReport.reportID === recentReport.reportID))
```
Comment (marcaaron) [inline src/pages/home/sidebar/SidebarLinks.js:180]: So we are taking the previous reports and replacing it with the new report... but why would the report itself be any different? Not sure if I understand what we are trying to achieve with this step in general. 

Though I do understand what we are doing after this to clean up the `undefined` reports.

File: src/pages/home/sidebar/SidebarLinks.js [hunk 5]
```diff
@@ -158,23 +158,35 @@ class SidebarLinks extends React.Component {
...
+            priorityMode: props.priorityMode,
             unreadReports: SidebarLinks.getUnreadReports(props.reports || {}),
         };
     }
 
     static getDerivedStateFromProps(nextProps, prevState) {
         const shouldReorder = SidebarLinks.shouldReorder(nextProps, prevState.orderedReports, prevState.currentlyViewedReportID, prevState.unreadReports);
+        const switchingPriorityModes = nextProps.priorityMode !== prevState.priorityMode;
+
+        // Pull the reports we want to show on the left hand nav, ordered by priority
```
Comment (marcaaron) [inline src/pages/home/sidebar/SidebarLinks.js:170]: This comment is a bit ambiguous. We are getting the formatted options to display from data that was already downloaded. Not sure how it was meant in this case.

**Review by marcaaron — DISMISSED** [review]
Changes look good, but I did have a question or two. I found this file really hard to follow and think it could be improved, but no expectations to do that in this PR. 😄

File: src/pages/home/sidebar/SidebarLinks.js (hunk 1 above)
Comment (ctkochan22) [reply src/pages/home/sidebar/SidebarLinks.js:168]: this.props isn't available within `getDerivedStateFromProps` for some reason. 

Out of curiosity, why do you want to leave state out of the equation? Bc we are not using `state.priorityMode` anywhere else?

File: src/pages/home/sidebar/SidebarLinks.js (hunk 3 above)
Comment (ctkochan22) [reply src/pages/home/sidebar/SidebarLinks.js:178]: I think we didn't want the lefthand nav to reorder themselves too frequently. Specifically when a user begins to type a message to a user (who might be halfway down the list). It won't immediately pop it to them to the top of the list, unless they leave a draft for that user. 

Honestly, I'm not entirely sure. I saw that you had authored a few thing in this file so I was hoping you'd have a strong idea of why haha .

Original Issue for lefthand re-order here: https://github.com/Expensify/App/issues/5166 with the PR here https://github.com/Expensify/App/pull/5317

File: src/pages/home/sidebar/SidebarLinks.js (hunk 4 above)
Comment (ctkochan22) [reply src/pages/home/sidebar/SidebarLinks.js:180]: So again, not super familiar with this area, but as you probably know each report object has info. Specifically for the draft messages. That report object's `hasDraftMessage` of the previous state will be set to false. The new list of reports pulled from `getRecentReports()` will have an updated list of report object, meaning if we were to start typing a message, the report object's `hasDraftMessage` key would be set to true.

File: src/pages/home/sidebar/SidebarLinks.js (hunk 1 above)
Comment (marcaaron) [reply src/pages/home/sidebar/SidebarLinks.js:168]: >Bc we are not using state.priorityMode anywhere else

That, and we're just setting whatever is in props directly to state which _usually_ means it does not need to be there. If we need to track the updates of a prop then we can just look at the prop.

Anyway, I think we are basically stuck at this point and would need to refactor the entire component to fix this. I'm remembering now that this is a _static_ method and can't access the component instance at all.

There are some good explanations here that go into why we should not use this here (but hopefully the explanation I just gave also makes sense).

https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html

File: src/pages/home/sidebar/SidebarLinks.js (hunk 2 above)
Comment (marcaaron) [reply src/pages/home/sidebar/SidebarLinks.js:189]: Never mind we can't do it without completely rewriting the component.

File: src/pages/home/sidebar/SidebarLinks.js (hunk 3 above)
Comment (marcaaron) [reply src/pages/home/sidebar/SidebarLinks.js:178]: Yeah, I did set up some things in the early days, but barely recognize it now.

File: src/pages/home/sidebar/SidebarLinks.js (hunk 4 above)
Comment (marcaaron) [reply src/pages/home/sidebar/SidebarLinks.js:180]: Ok so the because we are building a new set of options some things about those options will change. Makes sense, thanks.

File: src/pages/home/sidebar/SidebarLinks.js (hunk 5 above)
Comment (marcaaron) [reply src/pages/home/sidebar/SidebarLinks.js:170]: Maybe something like "Build the report options we want to show...."

File: src/pages/home/sidebar/SidebarLinks.js (hunk 1 above)
Comment (ctkochan22) [reply src/pages/home/sidebar/SidebarLinks.js:168]: I would actually be super interested in refactoring the entire component

However, should I scale back some of the changes then to get somethign merged now? Right now the undefineds are causing a crash, or requires a hard refresh

Comment (marcaaron) [reply src/pages/home/sidebar/SidebarLinks.js:168]: Yeah! I created this issue [here](https://github.com/Expensify/App/issues/7948) - feel free to steal it from me. 

But agree we should just fix this for now :D

**Comment by ctkochan22** [comment]
Updated. I tested it on all platforms, videos of the tests above. Definitely encouraged to test though! Added a lot of regression testing as well

**Comment by ctkochan22** [comment]
@pecanoro friendly bump!

**Review by pecanoro — APPROVED** [review]
